File: src/components/arx_fatalis/models/runes.rs
use std::fmt;
use strum_macros::EnumIter;

use crate::components::audio::{AudioOpt::*, AudioSrc};

#[derive(Debug, EnumIter)]
pub enum Rune {
    Aam,
    Nhi,
    Mega,
    Yok,
    Taar,
    Kaom,
    Vitae,
    Vista,
    Stregum,
    Morte,
    Cosum,
    Comunicatum,
    Movis,
    Tempus,
    Folgora,
    Spacium,
    Tera,
    Cetrius,
    Rhaa,
    Fridd,
}

impl fmt::Display for Rune {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}

impl Rune {
    pub fn as_src(&self) -> String {
        format!("/arx/runes/{}.mp3", self.to_string().to_lowercase())
    }
}

pub fn init_audio() -> Vec<AudioSrc> {
    vec![
        (Rune::Aam, [0.2, 0.7]),
        (Rune::Nhi, [0.15, 0.7]),
        (Rune::Mega, [0.15, 1.0]),
        (Rune::Yok, [0.2, 0.7]),
        (Rune::Taar, [0.25, 0.9]),
        (Rune::Kaom, [0.2, 1.05]),
        (Rune::Vitae, [0.2, 1.0]),
        (Rune::Vista, [0.3, 1.1]),
        (Rune::Stregum, [0.1, 1.25]),
        (Rune::Morte, [0.25, 1.2]),
        (Rune::Cosum, [0.15, 1.2]),
        (Rune::Comunicatum, [0.15, 1.65]),
        (Rune::Movis, [0.3, 1.2]),
        (Rune::Tempus, [0.3, 1.3]),
        (Rune::Folgora, [0.15, 1.2]),
        (Rune::Spacium, [0.3, 1.4]),
        (Rune::Tera, [0.2, 1.0]),
        (Rune::Cetrius, [0.15, 1.3]),
        (Rune::Rhaa, [0.3, 0.9]),
        (Rune::Fridd, [0.15, 1.5]),
    ]
    .into_iter()
    .map(|(rune, [start, next])| (rune.as_src(), vec![StartsAt(start), NextStartsAt(next)]))
    .collect()
}

File: src/types/arx-fatalis/runes.ts
export enum RuneEnum {
  aam,
  nhi,
  mega,
  yok,
  taar,
  kaom,
  vitae,
  vista,
  stregum,
  morte,
  cosum,
  comunicatum,
  movis,
  tempus,
  folgora,
  spacium,
  tera,
  cetrius,
  rhaa,
  fridd,
}

const RE = RuneEnum;

export class Rune {
  static arr: Rune[] = (() => {
    console.debug("initializing Runes array");

    return [
      [RE.aam, 0.2, 0.7],
      [RE.nhi, 0.15, 0.7],
      [RE.mega, 0.15, 1.0],
      [RE.yok, 0.2, 0.7],
      [RE.taar, 0.25, 0.9],
      [RE.kaom, 0.2, 1.05],
      [RE.vitae, 0.2, 1.0],
      [RE.vista, 0.3, 1.1],
      [RE.stregum, 0.1, 1.25],
      [RE.morte, 0.25, 1.2],
      [RE.cosum, 0.15, 1.2],
      [RE.comunicatum, 0.15, 1.65],
      [RE.movis, 0.3, 1.2],
      [RE.tempus, 0.3, 1.3],
      [RE.folgora, 0.15, 1.2],
      [RE.spacium, 0.3, 1.4],
      [RE.tera, 0.2, 1.0],
      [RE.cetrius, 0.15, 1.3],
      [RE.rhaa, 0.3, 0.9],
      [RE.fridd, 0.15, 1.5],
    ].map((row: any) => new Rune(row));
  })();

  static byVariant(variant: RuneEnum) {
    return this.arr.find((x) => x.variant === variant);
  }

  variant: RuneEnum;
  #startsAt: number;
  #length: number;
  #mp3: HTMLAudioElement;

  constructor([rune, start, stop]: [RuneEnum, number, number]) {
    this.variant = rune;
    const path = this.#url("mp3");
    this.#mp3 = new Audio(path);
    this.#mp3.preload = "auto";
    this.#startsAt = start;
    this.#length = Math.round((stop - start) * 1000);
  }

  play() {
    console.debug(`playing ${this.name}`);
    this.#mp3.currentTime = this.#startsAt;
    this.#mp3.play();
    return this.#length;
  }

  get name() {
    return this.str();
  }

  str(titleCase = true) {
    const name = RE[this.variant];

    if (titleCase) {
      const [initial, ...rest] = name;
      return [initial.toUpperCase(), ...rest].join("");
    }

    return name;
  }

  #url(ext = "mp3") {
    return `/arx/runes/${this.str(false)}.${ext}`;
  }

  get png() {
    return this.#url("png");
  }
}
